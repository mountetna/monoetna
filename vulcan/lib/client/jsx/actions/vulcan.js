export const SET_WORKFLOWS = 'SET_WORKFLOWS';
export const SET_WORKFLOW = 'SET_WORKFLOW';
export const SET_STATUS = 'SET_STATUS';
export const SET_DATA = 'SET_DATA';
export const SET_PATH = 'SET_PATH';
export const SET_STEP = 'SET_STEP';
export const SET_SESSION = 'SET_SESSION';
export const SET_INPUTS = 'SET_INPUTS';
export const SET_CALCULATING = 'SET_CALCULATING';
export const SET_DIRTY = 'SET_DIRTY';
