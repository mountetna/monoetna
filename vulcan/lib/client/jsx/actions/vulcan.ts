import {SessionStatusResponse, Workflow, WorkflowsResponse} from "../api_types";
import {InputValidator} from '../components/workflow/user_interactions/inputs/input_types';

function actionObject<T extends string, P>(type: T, payload: P): { type: T } & P {
    return { ...payload, type };
}

export function setWorkflows(workflows: WorkflowsResponse['workflows']) {
    return actionObject('SET_WORKFLOWS', {workflows});
}

export function setWorkflow(workflow: Workflow, projectName: string) {
    return actionObject('SET_WORKFLOW', {workflow, projectName});
}

export function setStatus(status: SessionStatusResponse['status']) {
    return actionObject('SET_STATUS', {status});
}

export function setDownloadedData(url: string, data: any) {
    return actionObject('SET_DOWNLOAD', {url, data});
}

export function releaseDownloadedData(url: string) {
    return actionObject('RELEASE_DOWNLOAD', {url});
}

export function setSession(session: SessionStatusResponse['session']) {
    return actionObject('SET_SESSION', {session});
}

// Fully replaces the inputs state.
export function setInputs(inputs: SessionStatusResponse['session']['inputs']) {
    return actionObject('SET_INPUTS', {inputs});
}

// Adds any inputs specified, but does not fully replace inputs.
export function patchInputs(inputs: SessionStatusResponse['session']['inputs']) {
    return actionObject('PATCH_INPUTS', {inputs});
}

// Removes the inputs by the given source name from the inputs and session states
export function removeInputs(inputs: string[]) {
    return actionObject('REMOVE_INPUTS', {inputs});
}

// Removes the downloads by the given step names from the status.  Usually this happens from server
// responses anyways, but it is a responsive UX feature.
export function removeDownloads(stepNames: string[]) {
    return actionObject('REMOVE_DOWNLOADS', {stepNames});
}

export function addValidationErrors(inputName: string, inputLabel: string, errors: string[]) {
    return actionObject('ADD_VALIDATION_ERRORS', {inputName, inputLabel, errors});
}

export function removeValidationErrors(inputName: string) {
    return actionObject('REMOVE_VALIDATION_ERRORS', {inputName});
}

export type VulcanAction = ReturnType<typeof setWorkflows> | ReturnType<typeof setWorkflow> | ReturnType<typeof setStatus> |
    ReturnType<typeof setDownloadedData> | ReturnType<typeof setSession> | ReturnType<typeof setInputs> |
    ReturnType<typeof releaseDownloadedData> | ReturnType<typeof removeInputs> |
    ReturnType<typeof patchInputs> | ReturnType<typeof removeDownloads> |
    ReturnType<typeof addValidationErrors> | ReturnType<typeof removeValidationErrors>;
