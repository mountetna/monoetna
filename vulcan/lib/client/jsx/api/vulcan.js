import * as _ from 'lodash';

import {
  checkStatus,
  handleFetchSuccess,
  handleFetchError,
  headers,
  isJSON
} from 'etna-js/utils/fetch';

const vulcanPath = (endpoint) => `${CONFIG.vulcan_host}${endpoint}`;

const vulcanPost = (endpoint, params) => {
  return fetch(endpoint, {
    method: 'POST',
    credentials: 'include',
    headers: headers('json'),
    body: JSON.stringify({
      ...params
    })
  }).then(checkStatus);
};

const rawVulcanGet = (endpoint) => {
  return fetch(endpoint, {
    method: 'GET',
    credentials: 'include',
    headers: headers('json')
  });
};

const vulcanGet = (endpoint) => {
  return rawVulcanGet(endpoint).then(checkStatus);
};

export const getWorkflows = () => {
  return vulcanGet(vulcanPath(ROUTES.fetch_workflows()))
    .then(handleFetchSuccess)
    .catch(handleFetchError);
};

// export const getWorkflow = (workflow_name, json = false) => {
//   // NOTE: response might be a YAML doc, for editing purposes.
//   // TODO: update this per real Vulcan endpoint
//   return vulcanGet(ROUTES.fetch_workflow(workflow_name, json))
//     .then(handleFetchSuccess)
//     .catch(handleFetchError);
// };

export const submit = (context) => {
  let {
    workflow,
    session,
    status,
    pathIndex,
    setSession,
    setStatus,
    setData
  } = context;

  let dataUrls = [];

  // Use a deep clone because we need to check if the download URL
  //   changed, later.
  let oldStatus = _.cloneDeep(status);

  return vulcanPost(vulcanPath(ROUTES.submit(workflow.name)), {
    inputs: session.inputs,
    key: session.key
  })
    .then(handleFetchSuccess)
    .then((response) => {
      setSession(response.session);
      setStatus(response.status);
      // Fetch data and update Context

      // Calculate this locally because useContext
      //   updates async?
      let updatedStatus = [...oldStatus].map((oldPath, oldPathIndex) => {
        return oldPath.map((oldStep, oldStepIndex) => {
          return {...oldStep, ...response.status[oldPathIndex][oldStepIndex]};
        });
      });

      let dataRequests = [];
      updatedStatus[pathIndex].forEach((step, stepIndex) => {
        if (step.downloads) {
          Object.keys(step.downloads).forEach((download) => {
            if (
              downloadUrlUpdated(
                oldStatus[pathIndex][stepIndex],
                step,
                download
              )
            ) {
              let dataUrl = step.downloads[download];
              dataUrls.push(dataUrl);
              dataRequests.push(
                // Don't want checkStatus here so that we
                //  can call Promise.all(), just the raw fetch.
                rawVulcanGet(dataUrl)
              );
            }
          });
        }
      });
      return Promise.all(dataRequests);
    })
    .then((downloads) => {
      let extractData = [];

      downloads.forEach((response) => {
        let data = isJSON(response) ? response.json() : response.text();

        extractData.push(data);
      });

      return Promise.all(extractData);
    })
    .then((extractions) => {
      extractions.forEach((datum, index) => {
        setData(dataUrls[index], datum);
      });
      return Promise.resolve();
    })
    .catch(handleFetchError);
};

// export for testing
export const downloadUrlUpdated = (oldStep, newStep, downloadKey) => {
  return (
    !(newStep.data && newStep.data[downloadKey]) ||
    newStep.downloads[downloadKey] !== oldStep.downloads[downloadKey]
  );
};

export const getSession = (workflow_name) => {
  // A "blank" POST to submit generates and returns the
  //   session, which we'll need for subsequent input
  //   submit actions.
  return vulcanPost(vulcanPath(ROUTES.submit(workflow_name)))
    .then(handleFetchSuccess)
    .catch(handleFetchError);
};

export const getData = (url) => {
  return vulcanGet(url).then(handleFetchSuccess).catch(handleFetchError);
};
