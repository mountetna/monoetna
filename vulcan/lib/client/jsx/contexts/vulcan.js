import React, {useReducer, createContext} from 'react';
import {
  SET_DATA,
  SET_WORKFLOW,
  SET_WORKFLOWS,
  SET_STATUS,
  SET_STEP,
  SET_PATH,
  SET_SESSION,
  SET_INPUTS
} from '../actions/vulcan';
import VulcanReducer from '../reducers/vulcan';

export const VulcanContext = createContext();

export const VulcanProvider = (props) => {
  const initialState = {
    workflows: {},
    workflow: {},
    pathIndex: null,
    stepIndex: null,
    session: {},
    status: {}
  };

  const [state, dispatch] = useReducer(VulcanReducer, initialState);

  const setWorkflows = (workflows) => {
    dispatch({type: SET_WORKFLOWS, workflows});
  };

  const setWorkflow = (workflow) => {
    dispatch({type: SET_WORKFLOW, workflow});
  };

  const setStatus = (status) => {
    dispatch({type: SET_STATUS, status});
  };

  const setData = (url, data) => {
    dispatch({type: SET_DATA, url, data});
  };

  const setPathIndex = (pathIndex) => {
    dispatch({type: SET_PATH, pathIndex});
  };

  const setStepIndex = (stepIndex) => {
    dispatch({type: SET_STEP, stepIndex});
  };

  const setSession = (session) => {
    dispatch({type: SET_SESSION, session});
  };

  const setInputs = (inputs) => {
    dispatch({type: SET_INPUTS, inputs});
  };

  return (
    <VulcanContext.Provider
      value={{
        workflows: state.workflows,
        workflow: state.workflow,
        pathIndex: state.pathIndex,
        stepIndex: state.stepIndex,
        session: state.session,
        setWorkflows,
        setWorkflow,
        setData,
        setStatus,
        setPathIndex,
        setStepIndex,
        setSession,
        setInputs
      }}
    >
      {props.children}
    </VulcanContext.Provider>
  );
};
