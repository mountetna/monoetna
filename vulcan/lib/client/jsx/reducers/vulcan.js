import {
  SET_DATA,
  SET_WORKFLOW,
  SET_WORKFLOWS,
  SET_STATUS,
  SET_PATH,
  SET_STEP,
  SET_SESSION,
  SET_INPUTS,
  SET_CALCULATING
} from '../actions/vulcan';

export default function VulcanReducer(state, action) {
  state = state ? state : {};

  switch (action.type) {
    case SET_WORKFLOWS:
      return {
        ...state,
        workflows: action.workflows
      };
    case SET_WORKFLOW:
      return {
        ...state,
        workflow: action.workflow
      };
    case SET_STATUS:
      // Update each entry in the status Arrays.
      // Workflow status are an Array of Arrays, so need to
      //   loop through them.
      // Assume stable order from the server.
      let currentStatus = [...(state.status || [])];
      action.status.forEach((path, pathIndex) => {
        if (!currentStatus[pathIndex]) currentStatus[pathIndex] = [];
        path.forEach((step, stepIndex) => {
          if (!currentStatus[pathIndex][stepIndex])
            currentStatus[pathIndex][stepIndex] = {};

          currentStatus[pathIndex][stepIndex] = {
            ...currentStatus[pathIndex][stepIndex],
            ...step
          };
        });
      });
      return {
        ...state,
        status: currentStatus
      };
    case SET_DATA:
      // Inject the data payload to status based on the URL.
      // Workflow status are an Array of Arrays, so need to
      //   loop through them to find matching data URLs.
      // Assume stable order from the server.
      let dataStatus = [...state.status];
      dataStatus.forEach((path, pathIndex) => {
        path.forEach((step, stepIndex) => {
          if (!step.downloads) return;

          Object.keys(step.downloads).forEach((downloadKey) => {
            if (action.url === step.downloads[downloadKey]) {
              if (!dataStatus[pathIndex][stepIndex].data) {
                dataStatus[pathIndex][stepIndex].data = {};
              }

              dataStatus[pathIndex][stepIndex].data[downloadKey] = action.data;
            }
          });
        });
      });

      return {
        ...state,
        status: dataStatus
      };
    case SET_PATH:
      return {
        ...state,
        pathIndex: action.pathIndex
      };
    case SET_STEP:
      return {
        ...state,
        stepIndex: action.stepIndex
      };
    case SET_SESSION:
      return {
        ...state,
        session: action.session
      };
    case SET_INPUTS:
      console.log('state', state);
      console.log('action', action);
      return {
        ...state,
        session: {
          ...(state.session || {}),
          inputs: {
            ...(state.session ? state.session.inputs : {}),
            ...action.inputs
          }
        }
      };
    case SET_CALCULATING:
      // This should go away once we have async jobs and
      //   a different way to check status.
      return {
        ...state,
        calculating: action.calculating
      };
    default:
      return state;
  }
}
