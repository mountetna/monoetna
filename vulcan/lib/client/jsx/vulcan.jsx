// Framework Libraries.
import * as React from 'react';
import * as ReactDOM from 'react-dom';

import {Provider} from 'react-redux';
import VulcanUI from './vulcan_ui';
import {VulcanStore} from './vulcan_store';
import * as Cookies from 'etna-js/utils/cookies';

class VulcanApplication {
  constructor(props, container_id) {
    // create the store
    this.createStore();

    // add user info from the token to the store
    this.store.dispatch({
      type: 'ADD_TOKEN_USER',
      token: Cookies.getItem(CONFIG.token_name)
    });

    // create the base component
    this.createUI(props, container_id);
  }

  createStore() {
    this.store = VulcanStore();
  }

  createUI({environment}, container_id) {
    ReactDOM.render(
      <Provider store={this.store}>
        <VulcanUI
          environment={environment}
          path={decodeURI(window.location.pathname)}
        />
      </Provider>,
      document.getElementById(container_id)
    );
  }
}

window.VulcanApp = VulcanApplication;
