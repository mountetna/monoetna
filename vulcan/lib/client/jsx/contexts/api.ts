import {useCallback, useRef, useState} from 'react';
import {showMessages} from 'etna-js/actions/message_actions';
import {
  checkStatus,
  handleFetchError,
  handleFetchSuccess,
  headers
} from 'etna-js/utils/fetch';
import {
  defaultSessionStatusResponse,
  defaultWorkflowsResponse,
  FiguresResponse,
  SessionStatusResponse,
  VulcanFigureSession,
  VulcanSession,
  WorkflowsResponse
} from '../api_types';

export const defaultApiHelpers = {
  showErrors<T>(work: Promise<T>): Promise<T> {
    work.catch((e) => {
      console.error(e);
    });

    return work;
  },
  getData(url: string): Promise<any> {
    return new Promise(() => null);
  },
  postInputs(session: VulcanSession): Promise<SessionStatusResponse> {
    return new Promise(() => null);
  },
  pollStatus(session: VulcanSession): Promise<SessionStatusResponse> {
    return new Promise(() => null);
  },
  getWorkflows(): Promise<WorkflowsResponse> {
    return new Promise(() => null);
  },
  fetchFigures(projectName: string): Promise<FiguresResponse> {
    return new Promise(() => null);
  },
  fetchFigure(
    projectName: string,
    figureId: number
  ): Promise<VulcanFigureSession> {
    return new Promise(() => null);
  },
  createFigure(projectName: string, params: any): Promise<VulcanFigureSession> {
    return new Promise(() => null);
  },
  updateFigure(projectName: string, params: any): Promise<VulcanFigureSession> {
    return new Promise(() => null);
  },
  deleteFigure(
    projectName: string,
    figureId: number
  ): Promise<VulcanFigureSession> {
    return new Promise(() => null);
  },
  updateFigureDependencies(projectName: string, figureId: number): Promise<VulcanFigureSession> {
    return new Promise(() => null);
  }
};

export function useApi(
  invoke: (a: {type: string}) => any
): typeof defaultApiHelpers {
  const vulcanPath = useCallback(
    (endpoint: string) => `${CONFIG.vulcan_host}${endpoint}`,
    []
  );
  const vulcanPost = useCallback((endpoint: string, params: Object) => {
    return fetch(endpoint, {
      method: 'POST',
      credentials: 'include',
      headers: headers('json'),
      body: JSON.stringify({
        ...params
      })
    }).then(checkStatus);
  }, []);

  const rawVulcanGet = useCallback((endpoint: string) => {
    return fetch(endpoint, {
      method: 'GET',
      credentials: 'include',
      headers: headers('json')
    });
  }, []);

  const vulcanDelete = useCallback((endpoint: string) => {
    return fetch(endpoint, {
      method: 'DELETE',
      credentials: 'include',
      headers: headers('json')
    });
  }, []);

  const vulcanGet = useCallback(
    (endpoint: string) => {
      return rawVulcanGet(endpoint)
        .then(checkStatus)
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [rawVulcanGet]
  );

  const getWorkflows = useCallback((): Promise<WorkflowsResponse> => {
    // old: ROUTES.fetch_workflows
    return vulcanGet(vulcanPath('/api/workflows'))
      .then(handleFetchSuccess)
      .catch(handleFetchError);
  }, [vulcanGet, vulcanPath]);

  const postInputs = useCallback(
    (session: VulcanSession): Promise<SessionStatusResponse> => {
      if (!session.workflow_name) {
        return Promise.reject(
          new Error('No workflow selected, bug in client.')
        );
      }

      return vulcanPost(
        // old: ROUTES.submit
        vulcanPath(`/api/${session.project_name}/session/${workflow_name}`),
        session
      )
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanPath, vulcanPost]
  );

  const pollStatus = useCallback(
    (session: VulcanSession): Promise<SessionStatusResponse> => {
      if (!session.workflow_name) {
        return Promise.reject(
          new Error('No workflow selected, bug in client.')
        );
      }

      return vulcanPost(
        // Old: ROUTES.status
        vulcanPath(`/api/${session.project_name}/session/${session.workflow_name}/status`),
        session
      )
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanPath, vulcanPost]
  );

  const getData = useCallback(
    (url: string) => {
      return vulcanGet(url)
        .then(handleFetchSuccess)
        .catch(handleFetchError)
        .then((data) => {
          // TODO: In the future, we should set content type headers to inform the client, for now we aggressively
          // try to parse JSON
          try {
            return JSON.parse(data);
          } catch {
            return data;
          }
        });
    },
    [vulcanGet]
  );

  const showErrors = useCallback(
    <T>(work: Promise<T>): Promise<T> => {
      work.catch((e) => {
        if (!(e instanceof Array)) {
          e = [`${e}`];
        }

        console.error(e);
        invoke(showMessages(e));
      });

      return work;
    },
    [invoke]
  );

  const fetchFigures = useCallback(
    (projectName: string): Promise<FiguresResponse> => {
      // Old: ROUTES.fetch_figures
      return vulcanGet(vulcanPath(`/api/${projectName}/figures`))
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanGet, vulcanPath]
  );

  const fetchFigure = useCallback(
    (projectName: string, figureId: number): Promise<VulcanFigureSession> => {
      // Old: ROUTES.fetch_figure
      return vulcanGet(vulcanPath(`/api/${projectName}/figure/${figureId}`))
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanGet, vulcanPath]
  );

  const updateFigure = useCallback(
    (projectName: string, params: any): Promise<VulcanFigureSession> => {
      return vulcanPost(
        // Old: ROUTES.update_figure
        vulcanPath(`/api/${projectName}/figure/${params.figure_id}/update`),
        params
      )
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanPost, vulcanPath]
  );

  const updateFigureDependencies = useCallback(
    (projectName: string, figure_id: number): Promise<VulcanFigureSession> => {
      return vulcanPost(
        // Old: ROUTES.update_figure
        vulcanPath(`/api/${projectName}/figure/${figure_id}/update`),
        {
          update_dependencies: true,
          comment: 'Updating dependencies and snapshot'
        }
      )
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanPost, vulcanPath]
  );

  const createFigure = useCallback(
    (projectName: string, params: any): Promise<VulcanFigureSession> => {
      // Old: ROUTES.create_figure
      return vulcanPost(vulcanPath(`/api/${projectName}/figure/create`), params)
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanPost, vulcanPath]
  );

  const deleteFigure = useCallback(
    (projectName: string, figureId: number): Promise<VulcanFigureSession> => {
      return vulcanDelete(
        // Old: ROUTES.delete_figure
        vulcanPath(`/api/${projectName}/figure/${figureId}`)
      )
        .then(handleFetchSuccess)
        .catch(handleFetchError);
    },
    [vulcanPath, vulcanDelete]
  );

  return {
    showErrors,
    getData,
    getWorkflows,
    postInputs,
    pollStatus,
    deleteFigure,
    createFigure,
    updateFigure,
    fetchFigures,
    fetchFigure,
    updateFigureDependencies
  };
}
