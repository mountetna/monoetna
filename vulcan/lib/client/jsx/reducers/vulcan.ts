import {
    SET_DATA,
    SET_WORKFLOW,
    SET_WORKFLOWS,
    SET_STATUS,
    SET_PATH,
    SET_STEP,
    SET_SESSION,
    SET_INPUTS,
    COMMIT_INPUTS, VulcanAction
} from '../actions/vulcan';
import {Workflow, WorkflowsResponse} from "../api/types";

export const defaultVulcanState = {
    workflows: {} as WorkflowsResponse['workflows'],
    workflow: null as (Workflow | null),

};

export type VulcanState = Readonly<(typeof defaultVulcanState)>;

export default function VulcanReducer(state: VulcanState, action: VulcanAction): VulcanState {
    state = state || defaultVulcanState;

    switch (action.type) {
        case 'SET_WORKFLOWS':
            return {
                ...state,
                workflows: action.workflows,
            };
        case 'SET_WORKFLOW':
            return {
                ...state,
                workflow: action.workflow
            };
        case SET_STATUS:
            // Update each entry in the status Arrays.
            // Workflow status are an Array of Arrays, so need to
            //   loop through them.
            // Assume stable order from the server.
            let currentStatus = [...(state.status || [])];
            action.status.forEach((path, pathIndex) => {
                if (!currentStatus[pathIndex]) currentStatus[pathIndex] = [];
                path.forEach((step, stepIndex) => {
                    if (!currentStatus[pathIndex][stepIndex])
                        currentStatus[pathIndex][stepIndex] = {};

                    currentStatus[pathIndex][stepIndex] = {
                        ...currentStatus[pathIndex][stepIndex],
                        ...step
                    };
                });
            });
            return {
                ...state,
                status: currentStatus
            };
        // case SET_DATA:
        //     // Inject the data payload to status based on the URL.
        //     // Workflow status are an Array of Arrays, so need to
        //     //   loop through them to find matching data URLs.
        //     // Assume stable order from the server.
        //     state = {...state};
        //     const dataCache = state.dataCache = Object.assign({}, state.dataCache);
        //     dataCache[action.url] = action.data;
        //     const dataStatus = state.status = [...state.status];
        //
        //     dataStatus.forEach((path, pathIndex) => {
        //         path.forEach((step, stepIndex) => {
        //             if (!step.downloads) return;
        //
        //             Object.keys(step.downloads).forEach((downloadKey) => {
        //                 if (action.url === step.downloads[downloadKey]) {
        //                     if (!dataStatus[pathIndex][stepIndex].data) {
        //                         dataStatus[pathIndex][stepIndex].data = {};
        //                     }
        //
        //                     dataStatus[pathIndex][stepIndex].data[downloadKey] = action.data;
        //                 }
        //             });
        //         });
        //     });
        //
        //     return state;
        // case SET_PATH:
        //     return {
        //         ...state,
        //         pathIndex: action.pathIndex
        //     };
        // case SET_STEP:
        //     return {
        //         ...state,
        //         stepIndex: action.stepIndex
        //     };
        // case SET_SESSION:
        //     return {
        //         ...state,
        //         session: action.session
        //     };
        // case SET_INPUTS:
        //     return {
        //         ...state,
        //         inputs: {
        //             ...(state.inputs || {}),
        //             ...action.inputs
        //         },
        //     };
        // case COMMIT_INPUTS:
        //     // Ensure the reference is shared, making equality comparison simpler.
        //     const newInputs = state.inputs || {};
        //
        //     return {
        //         ...state,
        //         inputs: newInputs,
        //         session: {
        //             inputs: newInputs,
        //         }
        //     };
        default:
            return state;
    }
}
