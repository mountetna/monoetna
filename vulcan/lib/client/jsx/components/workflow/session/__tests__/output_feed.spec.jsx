import React from 'react';
import {Provider} from 'react-redux';
import {mockStore} from 'etna-js/spec/helpers';
import {VulcanProvider} from '../../../../contexts/vulcan';
import renderer from 'react-test-renderer';
import OutputFeed from '../output_feed';

describe('OutputFeed', () => {
  let state;

  beforeEach(() => {
    state = {
      workflow: {
        inputs: {},
        steps: [
          [
            {
              name: 'first',
              run: 'scripts/something.cwl',
              out: ['output']
            },
            {
              name: 'second',
              run: 'ui-outputs/plotly.cwl',
              in: [{id: 'a', source: ['first', 'output']}],
              out: []
            },
            {
              name: 'third',
              run: 'scripts/make-link.cwl',
              in: [],
              out: ['output']
            },
            {
              name: 'fourth',
              run: 'ui-outputs/link.cwl',
              in: [{id: 'a', source: ['third', 'output']}],
              out: []
            },
            {
              name: 'fifth',
              run: 'scripts/make-raw.cwl',
              in: [],
              out: ['output']
            },
            {
              name: 'sixth',
              run: 'ui-outputs/raw.cwl',
              in: [{id: 'a', source: ['fifth', 'output']}],
              out: []
            }
          ]
        ]
      },
      pathIndex: 0,
      session: {
        key: 'session_key',
        inputs: {}
      },
      status: [
        [
          {
            name: 'first',
            status: 'complete',
            downloads: {
              output: 'https://foo'
            },
            data: {
              output: {data: [], layout: {}}
            }
          },
          {
            name: 'second',
            status: 'complete'
          },
          {
            name: 'third',
            status: 'complete',
            downloads: {
              output: 'https://foo'
            }
          },
          {
            name: 'fourth',
            status: 'complete'
          },
          {
            name: 'fifth',
            status: 'complete',
            downloads: {
              output: 'https://foo'
            },
            data: {
              output: 'some text!'
            }
          },
          {
            name: 'sixth',
            status: 'complete'
          }
        ]
      ]
    };
  });

  it('renders UI output steps', () => {
    // Link uses Redux store, so also include this Provider.
    const store = mockStore(state);
    // Wrap with Provider here so store gets passed down to child components in Context
    const component = renderer.create(
      <Provider store={store}>
        <VulcanProvider state={state}>
          <OutputFeed />
        </VulcanProvider>
      </Provider>
    );

    let instance = component.root;

    expect(instance.findAllByProps({className: 'step-output'}).length).toEqual(
      3
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
