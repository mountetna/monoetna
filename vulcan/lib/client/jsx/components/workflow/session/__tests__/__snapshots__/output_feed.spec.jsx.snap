// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutputFeed renders UI output steps 1`] = `
<div
  className="session-output-feed"
>
  <div
    className="step-output"
  >
    <div
      className="step-name"
    >
      <div
        className="step-status-icon-wrapper"
      >
        <span
          className="icon-grp fa-stack fa-fw"
        >
          <i
            className="icon step-status-icon light green check fa fa-2x fa-check"
          />
        </span>
      </div>
      <div
        className="step-button"
      >
        second
      </div>
    </div>
    <div
      className="outputs-pane"
    >
      <div
        style={
          Object {
            "display": "inline-block",
            "position": "relative",
          }
        }
      />
    </div>
  </div>
  <div
    className="step-output"
  >
    <div
      className="step-name"
    >
      <div
        className="step-status-icon-wrapper"
      >
        <span
          className="icon-grp fa-stack fa-fw"
        >
          <i
            className="icon step-status-icon light green check fa fa-2x fa-check"
          />
        </span>
      </div>
      <div
        className="step-button"
      >
        fourth
      </div>
    </div>
    <div
      className="outputs-pane"
    >
      <a
        className="link"
        href="https://foo"
        onClick={null}
      >
        Download data here
      </a>
    </div>
  </div>
  <div
    className="step-output"
  >
    <div
      className="step-name"
    >
      <div
        className="step-status-icon-wrapper"
      >
        <span
          className="icon-grp fa-stack fa-fw"
        >
          <i
            className="icon step-status-icon light green check fa fa-2x fa-check"
          />
        </span>
      </div>
      <div
        className="step-button"
      >
        sixth
      </div>
    </div>
    <div
      className="outputs-pane"
    >
      <div
        className="raw-view"
      >
        some text!
      </div>
    </div>
  </div>
</div>
`;
